new coalignment api (first draft) (#207)
sunpy/sunkit-image: docs/how_to_guide/adding_a_coalignment_method.rst

.. _sunkit-image-how-to-guide-adding-a-new-coalignment-method:

****************************
Add a New Coalignment Method
****************************

> You can add a custom coalignment method in the sunkit-image package using the decorator ``sunkit_image.coalignment_module.util.decorators.register_coalignment_method``:

Reviewer: These should be link able.

If they are not, then you will need to add files inside https://github.com/sunpy/sunkit-image/tree/main/docs/code_ref to enable that. Unfortunately, it's not automatic.

Author: I initially thought of referencing, but again since we would be replacing the current coalignment with the new one I thought it would make sense to avoid adding extra things and reference it later.

Reviewer: We should get them working, the renaming process won't be difficult. 

Should just be a find and replace.

.. code-block:: python

    from sunkit_image.coalignment_module.util.decorators import register_coalignment_method

    @register_coalignment_method("my_coalign")
    def my_coalignment_method(input_array, template_array):
        # Your coalignment code goes here
        # This should encompass calculating the shifts, applying these shifts to the data,
        # handling NaN values appropriately, and implementing any necessary clipping logic.
        # Return the shifts and the coaligned array
        return (x_shift, y_shift), coaligned_array

Decorator Parameters

> ====================
> - **"my_coalign"**: The name of your custom coalignment method.

Reviewer: ```suggestion
====================

- **"my_coalign"**: The name of your custom coalignment method.
```

I like a line break between heading and text

Function Requirements
=====================
Your coalignment function should:
1. **Take Input Parameters**:

    - ``input_array``: The 2D array to be coaligned.
    - ``template_array``: The 2D template array to align to.

2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``template_array``.

3. **Apply Shifts**: Apply these shifts to ``input_array`` to generate the coaligned array.

> 4. **Return**: A tuple where the first element is another tuple containing the shifts ``(x_shift, y_shift)``, and the second element is the coaligned array.

Reviewer: It should be the other way around, the array and then the shifts.

Author: Ah okay, I'll get it fixed

Example Usage
=============

> Once you have added your custom coalignment method, you can use it as discussed in this ``_sunkit-image-how-to-guide-using-the-coalignment-interface`` guide

Reviewer: These should updated to a :ref: link. It should work.

Author: yeah I'll add that.

> .. _add-a-new-coalignment-method:

Reviewer: These should be very specific as they were before.

We want to make sure that sphinx labels like this are not generic otherwise it can cause problems.

****************************
Add a New Coalignment Method
****************************

You can add a custom coalignment method in the sunkit-image package using the decorator :func:`sunkit_image.coalignment_module.register_coalignment_method`:

.. code-block:: python

    from sunkit_image.coalignment_module.util.decorators import register_coalignment_method

    @register_coalignment_method("my_coalign")
    def my_coalignment_method(input_array, template_array):
        # Your coalignment code goes here
        # This should encompass calculating the shifts, applying these shifts to the data,
        # handling NaN values appropriately, and implementing any necessary clipping logic.
        # Return the shifts and the coaligned array

>         return (x_shift, y_shift), coaligned_array

Reviewer: Array then shifts please.

Author: Ahh this was left out.

>         return (coaligned_array, (x_shift, y_shift))

Reviewer 2: Once we've firmed up what the return type of the method should be given our rethink this API, we should fix the documentation here to reflect that.

Decorator Parameters
====================

- **"my_coalign"**: The name of your custom coalignment method.

Function Requirements
=====================

Your coalignment function should:
1. **Take Input Parameters**:

    - ``input_array``: The 2D array to be coaligned.
    - ``template_array``: The 2D template array to align to.

2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``template_array``.

3. **Apply Shifts**: Apply these shifts to ``input_array`` to generate the coaligned array.

4. **Return**: A tuple where the first element is another tuple containing the shifts ``(x_shift, y_shift)``, and the second element is the coaligned array.

Example Usage
=============

> Once you have added your custom coalignment method, you can use it as discussed in this :doc:`using_new_coalignment_interface` guide

Reviewer: :ref: should work no?

Author: That actually didn't work. So I used doc, but I'll try that

Reviewer: Did it error? Do you recall what happened when you used ref?

Author: Yeah, it was unable to find the reference even after specifying the appropriate location.

Reviewer: That is odd, I wonder if there is a config issue in the package.

Author: > That is odd, I wonder if there is a config issue in the package.

Nope, issue was just the way I was referencing them. Its sorted I believe

> Once you have added your custom coalignment method, you can use it as discussed in :ref:`this <sunkit-image-how-to-guide-using-the-coalignment-interface>` guide

Reviewer: ```suggestion
Once you have added your custom coalignment method, you can use it as discussed in :ref:`sunkit-image-how-to-guide-using-the-coalignment-interface`.
```
Does this not work?

Author: It works, I just thought this would be better to read.

Reviewer: If I recall, using the ref means the title of the page shows up by default. I feel like that's more explicit than calling it this. But I need to check how it renders to make sure it looks ok.

.. _sunkit-image-how-to-guide-add-a-new-coalignment-method:

****************************
Add a New Coalignment Method
****************************

You can add a custom coalignment method in the sunkit-image package using the decorator :func:`~sunkit_image.utils.register_coalignment_method`:

.. code-block:: python

    from sunkit_image.utils.decorators import register_coalignment_method

    @register_coalignment_method("my_coalign")
    def my_coalignment_method(input_array, template_array):
        # Your coalignment code goes here
        # This should encompass calculating the shifts,
        # handling NaN values appropriately.

>         # Return the shifts in a affine style, such as the scale, rotation and translation.

Reviewer: ```suggestion
        # Return the shifts in an affine style, such as the scale, rotation and translation.
```

        return affine_params(scale, rotation, translation)

Decorator Parameters
====================

- **"my_coalign"**: The name of your custom coalignment method.

Function Requirements
=====================

Your coalignment function should:

1. **Take Input Parameters**:

   - ``input_array``: The 2D array to be coaligned.

>    - ``template_array``: The 2D template array to align to.

Reviewer: Should we not call this `target_array`?

2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``template_array``.

> 3. **Determine Affine Parameters**: Decide the parameters of the affine parameters like the scale, rotation and translation(generally shifts in x and y direction).

Reviewer: Are these not the same thing?

> 4. **Return**: A named tuple that contains the affine transformation parameters.

Reviewer: ```suggestion
4. **Return**: Use the ``affine_params`` named tuple from `sunkit_image` or provide your own that exposes the three parameters as attributes. 
```

> Example Usage
> =============
> 
> Once you have added your custom coalignment method, you can use it as discussed in :ref:`sunkit-image-how-to-guide-using-the-coalignment-interface`.

Reviewer: ```suggestion
```
Not sure we need this?

.. _sunkit-image-how-to-guide-add-a-new-coalignment-method:

****************************
Add a New Coalignment Method
****************************

> You can add a custom coalignment method in the sunkit-image package using the decorator :func:`~sunkit_image.coalignment.util.decorators.register_coalignment_method`:

Reviewer: ```suggestion
If you want to register a new coalignment method that can be used by :func:`~sunkit_image.coalignment.coalignment`, you can use :func:`~sunkit_image.coalignment.util.decorators.register_coalignment_method`:
```

.. code-block:: python

    from sunkit_image.utils.decorators import register_coalignment_method

    @register_coalignment_method("my_coalign")
    def my_coalignment_method(input_array, template_array):
        # Your coalignment code goes here
        # This should encompass calculating the shifts,
        # handling NaN values appropriately.
        # Return the shifts in an affine style, such as the scale, rotation and translation.
        return affine_params(scale, rotation, translation)

Decorator Parameters
====================

Currently the decorator takes one parameter:

> - ``name``: The name of your custom coalignment method, which in the above example is  "my_coalign"

Reviewer: ```suggestion
- ``name``: The name of your custom coalignment method, which in the above example is  "my_coalign".
```

Function Requirements
=====================

Your coalignment function should:

1. **Take Input Parameters**:

   - ``input_array``: The 2D array to be coaligned.

>    - ``template_array``: The 2D template array to align to.

Reviewer: ```suggestion
   - ``target_array``: The 2D array to align to.
```

> 2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``template_array``.

Reviewer: ```suggestion
2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``target_array``.
```

> 3. **Determine Affine Parameters**: Decide the parameters of the affine parameters like the scale, rotation and translation(generally shifts in x and y direction).

Reviewer: ```suggestion
3. **Determine Affine Parameters**: Decide the values of the affine parameters - scale, rotation and translation.
```

> 4. **Return**: Use the ``affine_params`` named tuple from `sunkit_image` or provide your own that exposes the three parameters as attributes.

Reviewer: ```suggestion
4. **Return**: Use the ``affine_params`` named tuple from `sunkit_image` or provide your own that exposes the three parameters as attributes.
```

> 4. **Return**: Use the ``affine_params`` named tuple from `sunkit_image` or provide your own that exposes the three parameters as attributes.

Reviewer: ```suggestion
4. **Return**: Use the ``affine_params`` named tuple included or provide your own that exposes the three parameters as attributes.
```

> 

Reviewer: ```suggestion

Handling NaNs and Infs
======================

Proper handling of these values is expected to be included in the registered methods.
The :func:`~sunkit_image.coalignment.coalignment` function does not change any problematic values. 
```

> Example Usage
> =============
> 
> Once you have added your custom coalignment method, you can use it as discussed in :ref:`sunkit-image-how-to-guide-using-the-coalignment-interface`.

Reviewer: ```suggestion
```

.. _sunkit-image-how-to-guide-add-a-new-coalignment-method:

****************************
Add a New Coalignment Method
****************************

If you want to register a new coalignment method that can be used by :func:`~sunkit_image.coalignment.coalign`, you can use :func:`~sunkit_image.coalignment.decorators.register_coalignment_method`:

.. code-block:: python

    from sunkit_image.coalignment.decorators import register_coalignment_method
    from sunkit_image.coalignment.interface import AffineParams

    @register_coalignment_method("my_coalign")

>     def my_coalignment_method(input_array, template_array):

Reviewer 3: ```suggestion
    def my_coalignment_method(input_array, target_array):
```

        # Return the shifts in an affine style, such as the scale, rotation and translation.
        return affine_params(scale, rotation, translation)

Decorator Parameters
====================

Currently the decorator takes one parameter:

- ``name``: The name of your custom coalignment method, which in the above example is  "my_coalign".

Function Requirements
=====================

Your coalignment function should:

1. **Take Input Parameters**:

   - ``input_array``: The 2D array to be coaligned.
   - ``target_array``: The 2D array to align to.

2. **Compute Shifts**: Calculate the shifts in the x and y directions needed to align ``input_array`` with ``target_array``.

3. **Determine Affine Parameters**: Decide the values of the affine parameters - translation, scale and rotation.

4. **Return**: Use the ``affine_params`` named tuple included or provide your own that exposes the three parameters as attributes.

Handling NaNs and Infs
======================

Proper handling of these values is expected to be included in the registered methods.
The :func:`~sunkit_image.coalignment.coalignment` function does not change any problematic values.

> 

Reviewer: ```suggestion
```

Checking if the Method is Registered

> ====================================
> To check if your method is registered, you can check if it is present in the registered methods dictionary :func:`~sunkit_image.coalignment.decorators.registered_methods` using the following code:

Reviewer: ```suggestion
====================================

To check if your method is registered, you can check if it is present in the registered methods dictionary :func:`~sunkit_image.coalignment.decorators.registered_methods` using the following code:
```

.. code-block:: python

>     import pprint

Reviewer: What is the difference in output from print vs pprint?

Author: there is no actual difference in the output, just thought this might be better. Simple ftw
